<style>
     p {
         max-width: 700px;
         text-align: left;
     }
     ul {
         max-width: 700px;
         text-align: left;
     }
     ul li {
         margin-bottom: 0px;
         padding: 5px;
     }   

     body {
       font-family: Verdana, sans-serif;
       font-size: 13px;
       width: 100%;
       max-width: 700px;
       margin: 30px;
       display: block;
       text-align: left;
     }
     .frame {

         width: 100%;
         border: 0px;
         white-space: nowrap;
         text-align: left; 
         margin: 1em 0;
     }
     .helper {
         display: inline-block;
         height: 100%;
         vertical-align: top;
     }
     img {
         background: #FFFFFF;
         vertical-align: top;
     }

     table {
         border-collapse: collapse;
         width: 100%;
     }
     th.date {
         width: 20%
     }
     th.subject {
         width: 70%; /* Not necessary, since only 70% width remains */
     }

     tr {
         border-bottom: 1px solid #ccc;
     }

     th {
         text-align: left;    
     }
     th, td {
         padding: 5px;
     }
</style>
<h1 id="tidecluster-report">TideCluster report</h1>
<h2 id="analysis-settings">Analysis settings:</h2>
<pre><code>SETTINGS_PLACEHOLDER</code></pre>
<h2 id="summary">Summary</h2>
<pre><code>SUMMARY_PLACEHOLDER</code></pre>
<h2 id="tandem-repeat-analysis-tarean">Tandem Repeat Analysis (TAREAN)</h2>
<p><a href="PREFIX_PLACEHOLDER_tarean_report.html"><strong>Report</strong></a></p>
<p>Report includes TAREAN analysis for all tandem repeat clusters (TRCs) pass the threshold of the minimamal combined length of tandem repeat arrays within a single cluster</p>
<ul>
<li><strong>Consensus sequence</strong> for each tandem repeat cluster (TRC) estimated by k-mer-based de Brujn graph approach.</li>
<li><strong>Consensus sequence for simple sequence repeats</strong> (SSRs) estimated by simplified k-mer-based approach.</li>
<li>Basic <strong>TRC statistics</strong> (number of arrays, total length, minimal and maximal length of arrays)</li>
</ul>
<h2 id="k-mer-interval-tandem-repeat-estimation-kite">K-Mer Interval Tandem Repeat Estimation (KITE)</h2>
<p><a href="PREFIX_PLACEHOLDER_kite_report.html"><strong>Report</strong></a></p>
<p>The K-Mer Interval Tandem Repeat Estimation (KITE) method is applied to analyze individual Tandem Repeat Arrays (TRAs) within each TRC. In contrast to TAREAN, which provides monomer size estimates for the entire TRC, KITE also estimates monomer sizes for each array by evaluating the distances between all repetitive k-mers across the tandem repeat array, with each tandem array analyzed individually. This method facilitates the detection of higher-order repeats and captures the variability in monomer size across different tandem repeat arrays. The KITE analysis includes:</p>
<ul>
<li><strong>Monomer size estimate for the entire TRC</strong> : This estimate may differ from the TAREAN estimate due to the different methodologies employed.</li>
<li><strong>Monomer size estimates for individual TRAs</strong> : These estimates are provided as:
<ul>
<li><strong>Primary estimate</strong> : The estimate with the highest score.</li>
<li><strong>Alternative estimates</strong> : Estimates with lower scores, these estimates help in identifying <strong>higher-order repeats</strong>.</li>
</ul></li>
</ul>
<h2 id="trc-consensus-clustering-and-superfamily-analysis">TRC Consensus Clustering and Superfamily Analysis</h2>
<p><a href="PREFIX_PLACEHOLDER_trc_similarity.html"><strong>Report</strong></a></p>
<p>Consensus sequences from TAREAN are clustered to identify similary or nearly identical TRCs tha were not clustered together initially in clustering step. This clustering groups the TRCs into higher-level clusters referred to as <strong>Superfamilies</strong>, based on sequence similarity.</p>
<p>Report includes:</p>
<ul>
<li>Dotplots of TAREAN Consensus Sequences: Visual representations showing pairwise sequence similarities among TRCs within each Superfamily.</li>
<li>Superfamily Groupings: Summaries listing the TRCs included in each Superfamily</li>
</ul>
<h2 id="credits">Credits</h2>
<p>TideCluster utilizes Tidehunter [https://github.com/Xinglab/TideHunter] for tandem repeat detection and TAREAN for reconstruction of consensus sequences of tandem repeats. If you use TideCluster please cite:</p>
<ul>
<li>https://github.com/kavonrtep/TideCluster <a href="https://zenodo.org/badge/latestdoi/601111441"><img src="https://zenodo.org/badge/601111441.svg" alt="DOI" /></a></li>
<li>TAREAN: a computational tool for identification and characterization of satellite DNA from unassembled short reads (https://doi.org/10.1093/nar/gkx257)</li>
<li>TideHunter: efficient and sensitive tandem repeat detection from noisy long-reads using seed-and-chain (https://doi.org/10.1093/bioinformatics/btz376)</li>
</ul>
<h3 id="authors">Authors</h3>
<p>Petr Novak, Jiri Macas, Laboratory of Molecular Cytogenetics, Biology Centre CAS, Czech Republic</p>
